File: java/src/main/java/com/leetcode/FourOneTwo.java
package com.leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author: John
 * @date: 2020-02-29 18:09
 * @description: 412
 **/
public class FourOneTwo {
    public List<String> fizBuzz(int n) {
        // avoid % to improvement performance
        List<String> ret = new ArrayList<>();
        for (int i = 1, fizz = 0, buzz = 0; i <= n; i++) {
            fizz++;
            buzz++;
            if (fizz == 3 && buzz == 5) {
                ret.add("FizzBuzz");
                fizz = 0;
                buzz = 0;
            } else if (fizz == 3) {
                ret.add("Fizz");
                fizz = 0;
            } else if (buzz == 5) {
                ret.add("Buzz");
                buzz = 0;
            } else {
                ret.add(String.valueOf(i));
            }
        }
        return ret;
    }
}

File: leetcodecpp/leetcode/FourOneTwo.cpp
//
// Created by John on 2020/2/29.
//
#include "ostream"
#include "vector"

using namespace std;

class Solution{
public:
//    vector<string> fizzBuzz(int n) {
//        vector<string> res;
//        for (int i = 1; i < n + 1; i++) {
//            if (i % 3 == 0 && i % 5 == 0) {
//                res.push_back("FizzBuzz");
//                continue;
//            }
//            if (i % 3 == 0) {
//                res.push_back("Fizz");
//                continue;
//            }
//            if (i % 5 == 0) {
//                res.push_back("Buzz");
//                continue;
//            }
//            res.push_back(to_string(i));
//        }
//        return res;
//    }
    vector<string> fizzBuzz(int n) {
        vector<string> res;
        for (int i = 1; i < n + 1; i++) {
            if (i % 3 == 0 && i % 5 == 0) {
                res.push_back("FizzBuzz");
                continue;
            }
            if (i % 3 == 0) {
                res.push_back("Fizz");
                continue;
            }
            if (i % 5 == 0) {
                res.push_back("Buzz");
                continue;
            }
            res.push_back(to_string(i));
        }
        return res;
    }
};

//int main() {
//    Solution * solution = new Solution();
//    solution->fizzBuzz(15);
//}
